2.md

```diff
@@ -0,0 +1,273 @@
+accomplished with a small virtual module source.
+
+```js
+class JsonModuleSource {
+  bindings = { export: 'default' };
+  constructor(text) {
+    this.value = JSON.parse(text);
+  };
+  execute(imports) {
+    imports.default = this.value;
+  };
+}
+
+const source = new JsonModuleSource({ meaning: 42 });
+const module = new Module(source);
+const { default: { meaning } } = await import(module);
+```
+
+Asset modules of various kinds would largely follow this pattern.
+
+### WASM
+
+On a host that does not provide support for WebAssembly, a virtual module
+source would suffice.
+
+```js
+class WasmModuleSource {
+  constructor(buffer) {
+    const module = new WebAssembly.Module(buffer);
+    this.#imports = WebAssembly.Module.imports(module);
+    this.#exports = WebAssembly.Module.exports(module);
+    this.bindings = [
+       ...this.#imports.map(({ module, name }) =>
+         ({ import: name, from: module })),
+       ...this.#exports.map(({ name }) =>
+         ({ export: name })),
+    ];
+  };
+  async execute(namespace) {
+    const importObject = {};
+    for (const { module, name, kind } of this.#imports) {
+      importObject[name] = namespace[name];
+    }
+    const instance = await WebAssembly.instantiate(module, importObject);
+    for (const { name } of this.#exports) {
+      namespace[name] = instance[name];
+    }
+  };
+}
+```
+
+### Pass-through module sources
+
+This example illustrates how a virtual module source can simply
+reexport another module with no special logic in an executor.
+
+```js
+import * as direct from 'real.js';
+const source = { bindings: { exportAllFrom: 'real.js', as: 'real' } };
+const module = new Module(source);
+const { real: indirect } = await import(module);
+direct === indirect; // true
+```
+
+## Dependencies
+
+This proposal depends on [Module and ModuleSource](0.md) from the [Compartments
+proposal](README.md) to introduce `ModuleSource`, and [ModuleSource
+analysis](1.md) from the [Compartments proposal](README.md).
+
+## Design
+
+For every `Module` instance that has a virtual module source,
+module import machinery constructs a real ***Module Imports Namespace***,
+an exotic object that allows user code to get and set the values for each
+binding.
+
+Bindings must be one of the shapes proposed in [module source static
+analysis](1.md), such that for each binding gets linked in a virtual module:
+
+- `{ import: string, from: string }`
+
+  Links the named `import` property of the ***Module Imports Namespace***
+  to the same name of the `from` module's ***Module Exports Namespace***.
+
+- `{ import: string, as: string, from: string }`
+
+  Links the named `as` property of the ***Module Imports Namespace***
+  to the `import` name of the `from` module's ***Module Exports Namespace***.
+
+- `{ export: string }`
+
+  Links the named `export` property of this module's ***Module Imports
+  Namespace*** directly to its ***Module Exports Namespace***.
+
+- `{ export: string, as: string}`
```

Comment:
What's the binding of `export const a = 1; export { a as b }`?

Reply:
Should be `{ export: 'a', as: 'b' }`. The executor is responsible for enforcing the constness of the module imports namespace.

Reply:
but `a` does not in the module's Module Exports Namespace.

Reply:
I agree, we only need `as` for the `export { a as b } from "x"` syntax.

Reply:
I agree, we only need `as` for the `export { a as b } from "x"` syntax. If it's exporting a variable declared in the same module, from the outside it's identical to `export const a = 1; export const b = a;`

Reply:
Thank you both. Good catch!

Reply:
> from the outside it's identical to export const a = 1; export const b = a;

If it is a mutable binding, it will not be identical (live binding), but it should be taken care of by the virtual module itself (update both bindings when assignment). There's no need to add support for it.